Import/port core-text Chinese composition tables

vitorroriz opened the pull request:
#### af11c789096b362380e9b0a32e96f5b00c25110a
<pre>
Import/port core-text Chinese composition tables
<a href="https://bugs.webkit.org/show_bug.cgi?id=252171">https://bugs.webkit.org/show_bug.cgi?id=252171</a>
rdar://105386285

Reviewed by Brent Fulgham.

This is a preparation for text-spacing implementation.
We need to import CoreText&apos;s tables and types to WebKit to mimic
CoreText&apos;s behavior when text-spacing is set to ‘auto’ for Chinese text.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.cpp: Added.
(characterToCharacterClass):
(isDigit):
(isNarrowFullWidth):
(ChineseCompositionRules::characterClass):
(ChineseCompositionRules::characterSpacing):
* Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.h: Added.
* Source/WebCore/platform/text/mac/CoreTextCompositionEngine.cpp: Added.
(CompositionRules::characterSpacing):
* Source/WebCore/platform/text/mac/CoreTextCompositionEngine.h: Added.

Canonical link: <a href="https://commits.webkit.org/260534@main">https://commits.webkit.org/260534@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3233705a414934650b33a7ccf82ad96c8e181cd2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108358 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41223 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117474 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112245 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18957 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8732 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100581 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114126 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14212 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97398 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42121 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29043 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/83801 "Found 1 new API test failure: /TestWebKit:WebKit.PrivateBrowsingPushStateNoHistoryCallback (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10285 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30387 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11027 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7291 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16435 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49984 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7280 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12614 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.h:
```diff
@@ -0,0 +1,55 @@
+class TChineseCompositionEngine {
+public:
+    // static bool CanHangRight(CTCompositionLanguage, bool, UTF32Char);
+    // static CGFloat GetLeftHangingFactor(CTCompositionLanguage, bool, UTF32Char);
```
If you plan to implement these two methods, I'd prefer you have the uncommented here, and just provide a stub with a `notImplemented()` call in the body.

vitorroriz replied:
ops, I forgot to remove these. Thanks.

brentfulgham replied:
Seems like the whole class should go away then?

brentfulgham commented on Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.cpp:
```diff
@@ -0,0 +1,191 @@
+        return 0;
+
+    /* Characters with the UCHAR_EAST_ASIAN_WIDTH enumerable property equal to U_EA_FULLWIDTH or U_EA_WIDE are of width 2. */
+    int width = (int)u_getIntPropertyValue(c, UCHAR_EAST_ASIAN_WIDTH);
```
Nit (style): We should use a C++ cast here: probably `static_cast<int>(u_getIntProperty....`

brentfulgham commented on Source/WebCore/platform/text/mac/CoreTextCompositionEngine.cpp:
```diff
@@ -0,0 +1,39 @@
+
+CharacterSpacingType GetCharacterSpacing(CTCompositionLanguage language, bool isVertical, UTF32Char before, CharacterClass beforeClass, UTF32Char after, uint32_t afterGeneralCategoryMask, CharacterClass* afterClass)
+{
+    // FIXME: add support for other languages.
```
Nit: When using a FIXME comment, we prefer to provide a bugzilla # (e.g., `// FIXME(123456): add support for other languages.`)

brentfulgham reviewed (approved):
LGTM, with a couple of small nits.

litherum commented on Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.cpp:
```diff
@@ -0,0 +1,191 @@
+    return ch == 0x6708 /* 月 */ || ch == 0x65E5 /* 日 */;
+}
+
+auto ChineseCompositionRules::GetCharacterClass(UTF32Char ch, uint32_t gc, CTCompositionLanguage language) -> CharacterClass
```
WebKit style is to use camelCase for function names.

vitorroriz committed:
Import/port core-text Chinese composition tables
https://bugs.webkit.org/show_bug.cgi?id=252171
rdar://105386285

Reviewed by Brent Fulgham.

This is a preparation for text-spacing implementation.
We need to import CoreText's tables and types to WebKit to mimic
CoreText's behavior when text-spacing is set to ‘auto’ for Chinese text.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.cpp: Added.
(characterToCharacterClass):
(isDigit):
(isNarrowFullWidth):
(ChineseCompositionRules::characterClass):
(ChineseCompositionRules::characterSpacing):
* Source/WebCore/platform/text/mac/CoreTextChineseCompositionEngine.h: Added.
* Source/WebCore/platform/text/mac/CoreTextCompositionEngine.cpp: Added.
(CompositionRules::characterSpacing):
* Source/WebCore/platform/text/mac/CoreTextCompositionEngine.h: Added.

Canonical link: https://commits.webkit.org/260534@main
